tt_metal/llrt/llrt.cpp

```diff
@@ -163,19 +163,27 @@ void write_launch_msg_to_core(chip_id_t chip, const CoreCoord core, launch_msg_t
         assert(is_active_eth_core or is_inactive_eth_core);
     }
 
-    msg->mode = DISPATCH_MODE_HOST;
+    msg->kernel_config.mode = DISPATCH_MODE_HOST;
+    uint32_t launch_addr;
+    uint32_t go_addr;
     if (is_active_eth_core) {
-        tt::Cluster::instance().write_core(
-            (void *)msg, sizeof(launch_msg_t), tt_cxy_pair(chip, core), GET_ETH_MAILBOX_ADDRESS_HOST(launch));
+        launch_addr = GET_ETH_MAILBOX_ADDRESS_HOST(launch.kernel_config);
+        go_addr = GET_ETH_MAILBOX_ADDRESS_HOST(launch.go);
     } else {
         if (is_inactive_eth_core) {
-            tt::Cluster::instance().write_core(
-                (void *)msg, sizeof(launch_msg_t), tt_cxy_pair(chip, core), GET_IERISC_MAILBOX_ADDRESS_HOST(launch));
+            launch_addr = GET_IERISC_MAILBOX_ADDRESS_HOST(launch.kernel_config);
+            go_addr = GET_IERISC_MAILBOX_ADDRESS_HOST(launch.go);
         } else {
-            tt::Cluster::instance().write_core(
-                (void *)msg, sizeof(launch_msg_t), tt_cxy_pair(chip, core), GET_MAILBOX_ADDRESS_HOST(launch));
+            launch_addr = GET_MAILBOX_ADDRESS_HOST(launch.kernel_config);
+            go_addr = GET_MAILBOX_ADDRESS_HOST(launch.go);
         }
     }
+
+    tt::Cluster::instance().write_core((void *)&msg->kernel_config, sizeof(kernel_config_msg_t), tt_cxy_pair(chip, core), launch_addr);
+    tt_driver_atomics::sfence();
+    if (send_go) {
```

Comment:
Before changes, we always wrote the go msg as well? Not writing it is a bug fix?

Reply:
latent bug - since run was in the same cacheline as exist_erisc_kernel, there was no race.  w/ my changes, these were written in separate noc transactions to raced w/ the kernel.  now, run is not written when we don't want to run.